File: logicsim.cpp
#include<bits/stdc++.h>
using namespace std;

int and_op(vector<int> x){
    int ans = 1;
    for(auto i:x){
        ans &= i;
    }
    return ans;
}
int or_op(vector<int> x){
    int ans = 0;
    for(auto i: x){
        ans |= i;
    }
    return ans;
}
int not_op(int x){
    int ans = ~x;
    return ans;
}
void dfs(int node, vector<int> &v, vector<vector<int>> &adj, vector<bool> &visited, vector<char> &funcs){
    if(funcs[node] == 'i')
        return;
    if(visited[node])
        return;
    visited[node] = true;
    vector<int> inputs;
    for(auto i: adj[node]){
        dfs(i, v, adj, visited, funcs);
        inputs.push_back(v[i]);
    }
    switch(funcs[node]){
    case 'o':
        v[node] = or_op(inputs);
        break;
    case 'a':
        v[node] = and_op(inputs);
        break;
    case 'n':
        v[node] = not_op(inputs[0]);
        break;
    default:
        cout<<"Incorrect character entered"<<endl;
    }
    return;
}

int main()
{
    int n, m;
    cin>>n>>m;
    vector<int> v(n+m);
    vector<char> funcs(n+m);
    for(int i=0; i<m; i++){
        cin>>v[i];
        if((v[i] > 1) || (v[i]<0)){
            cout<<"Gaand maraiye Viru ji";
            return 0;
        }
        funcs[i] = 'i';
    }
    vector<vector<int>> adj(n+m,vector<int>());
    for(int i=0; i<m; i++){
        char ch;
        cin>>ch;
        int temp_n;
        cin>>temp_n;
        int temp;
        cin>>temp;
        for(int j=0; j<temp_n; j++){
            int x;
            cin>>x;
            adj[temp].push_back(x);
        }
        funcs[temp] = ch;
    }
    vector<bool> visited(m+n, false);
    for(int i=n+m-1; i>=0; i--){
        if(!visited[i])
            dfs(i, v, adj, visited, funcs);
    }
    int num_out;
    cin>>num_out;
    vector<int> outputs(num_out);
    for(int i=0; i<num_out; i++){
        cin>>outputs[i];
    }
    for(int i=0; i<num_out; i++){
        cout<<outputs[i];
    }
    return 0;
}

File: logicsim.py
#19D070029---Kartikeya Chandra:
#defining gates operations
def and_gat(x):
    ans = 1
    for i in x:
        ans = (ans and i)
    return ans

def or_gat(x):
    ans = 0
    for i in x:
        ans = (ans or i)
    return ans

def not_gate(x):
    ans = int(not x)
    return ans

def dfs(node, vect, adj, stop, funcs):
    if((funcs[node]=='i') or (stop[node])):
        return
    stop[node] = True
    inputs = []
    for i in adj[node]:
        dfs(i, vect, adj, stop, funcs)
        inputs.append(vect[i])
    if(funcs[node] == 'or'):
        vect[node] = or_gat(inputs)
    elif(funcs[node] == 'and'):
        vect[node] = and_gat(inputs)
    elif(funcs[node] == 'not'):
        vect[node] = not_gate(inputs[0])
    else:
        print("Incorrect character entered")
    return

(n,m) = map(int,input().split())
funcs = ['i' for i in range(m+n)]
vect = [int(i) for i in list(input())] + [0 for i in range(m)]
adj = [[] for i in range(m+n)]
inp_taken = [True for i in range(n)] + [False for i in range(m)]
for i in range(m):
    x = input().split()
    node = int(x[2])
    if inp_taken[node]:
        print('Incorrect Input givecten, pls try again')
        break
    inp_taken[node] = True
    funcs[node] = x[0]
    adj[node] = [int(i) for i in x[3:]]
stop = [False for i in range(n+m)]
for i in range(n+m-1,-1,-1):
    dfs(i,vect,adj,stop,funcs)
out_num = int(input())
nodes_out = [int(i) for i in input().split()]
for i in nodes_out:
    print(vect[i], end='')
